added option to use custom graphite prefix

christiangroth opened the pull request:
It would be great to be able to configure a custom prefix for graphite metrics. I'm running more than one ninja app on the same host and thus need to keep graphite metrics separated. This pull request is a simple solution to prefix each application and does not change anything if no configuration is made at all, using the hostname as fallback as it was before.

christiangroth committed:
added option to use custom graphite prefix

raphaelbauer commented:
Looks good :)

It would be great if you could test that with a mocked test - just to make sure no-one destroys your shiny feature accidentally...

And it would be great if you could add some documentation (metrics.md). Let's tell people about the option :)

christiangroth committed:
- downgrade unit to be able to run with power mock version
- added docs for custom graphite prefix
- added mocked test for custom graphite prefix

christiangroth commented:
Done! :)

Unfortunately I had to downgrade junit from 4.12 to 4.11 to be able to get powermock running at all. Needed this to mock the static methods getting invoked. Just waiting for travis CI to check it, my local build was green.

raphaelbauer commented:
👍 Cool. I'll have a look in the coming days :)

raphaelbauer commented on ninja-metrics-graphite/src/test/java/ninja/metrics/graphite/NinjaGraphiteTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    private void mockPrefix(String prefix) {
+        Mockito.when(ninjaProperties.get(Mockito.eq("metrics.graphite.prefix"))).thenReturn(prefix);
```
Please add this inside the test methods and don't use a separate method. Usually I'd agree that a well named method makes a lot of sense - but because it is just one line it is better imho to move it directly into the test methods.

raphaelbauer commented on ninja-metrics-graphite/src/test/java/ninja/metrics/graphite/NinjaGraphiteTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+
```
One newline too much

raphaelbauer commented on ninja-metrics-graphite/src/test/java/ninja/metrics/graphite/NinjaGraphiteTest.java:
```diff
@@ -0,0 +1,93 @@
+import ninja.utils.NinjaPropertiesImpl;
+
+@PrepareForTest(GraphiteReporter.class)
+@RunWith(PowerMockRunner.class)
```
I hate PowerMock - but in that case it makes total sense :)

raphaelbauer commented on pom.xml:
```diff
@@ -830,7 +830,7 @@
                 <groupId>junit</groupId>
                 <artifactId>junit</artifactId>
-                <version>4.12</version>
+                <version>4.11</version>
```
Please don't downgrade, but do the following:

- Keep Junit at 4.12
- Upgrade mockito-core to 1.10.19
- Upgrade powermock.version to 1.6.5

Then fix two lines in test 
ninja-maven-plugin/src/test/java/ninja/build/WatchAndRestartMachineTest.java (you'll get exceptions anyway when building).

```
verify(restartTrigger, timeout(10000).never()).trigger(); 
=>
verify(restartTrigger, Mockito.after(10000).never()).trigger();
```

Then we got the latest versions and everything is fine :)

raphaelbauer commented on ninja-metrics-graphite/src/test/java/ninja/metrics/graphite/NinjaGraphiteTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    public void testWithoutCustomPrefix() {
```
Nice testcase! Well done :)

raphaelbauer commented:
Hi @christiangroth - well done and nice tests and documentation. I added some comments regarding the junit downgrade (which we should not do). Easy to fix I hope :)

Apart from that please do the following:
- Sign https://docs.google.com/forms/d/e/1FAIpQLSd_I57jBuXkfoJffIjFyPtdxA3hq9OIr0c08gplpJnmAuxeTQ/viewform
- Add one line to changelog.md
- And your name to team.md.

Then we can merge it asap :)
Thanks for your contribution. Well appreciated 👍

christiangroth committed:
refactored test

christiangroth committed:
powermock and mockito upgrade

christiangroth committed:
team and changelog update

christiangroth commented:
Thanks for your feedback. I've pushed the changes. :)

raphaelbauer commented on ninja-maven-plugin/src/test/java/ninja/build/WatchAndRestartMachineTest.java:
```diff
@@ -1,88 +1,74 @@
 /**
- * Copyright (C) 2012-2017 the original author or authors.
- *
```
Can you please undo all the changes that are not related to the two lines you changed? Otherwise we are having a hard time figuring out what happened later... Thanks!

raphaelbauer commented:
Looks almost perfect - I added one tiny comment regarding the testcase for the watcher... Then we can merge it :)

christiangroth committed:
Revert "powermock and mockito upgrade"

This reverts commit 3e50243c60218bf486a885194f801feb5e800f8c.

christiangroth committed:
powermock and mockito upgrade

christiangroth committed:
fix indentation

christiangroth commented:
Sorry, there has happened some code formatter madness. Fixed this!!

raphaelbauer commented:
👍
